### PR TITLE
fix(UI): Script displayed in preview when using specific css

### LINKED ISSUE #55861: Script displayed in preview when using specific CSS
### Describe the Issue

Someone on the forum reported unexpected behaviour from the fcc editor here:
https://forum.freecodecamp.org/t/cash-register-cryptic-error-in-preview-screen/705785

Can someone take a look because it is not clear whether it is a bug or a limitation?

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/build-a-cash-register-project/build-a-cash-register

### Your code

```

* {
  margin: 0;
  padding: 0;
  display: block;
}

```

### Expected behavior

CSS should not cause errors in editor

### Screenshots

_No response_

### System

Laptop/windows/chrome

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Othniel01**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #55861 

This pull request introduces enhancements to the createHeader function, which now injects CSS to ensure that specific head elements (including title, meta, link, and script tags) are hidden in the preview frame. This change addresses a previously identified issue where these elements were being rendered in the preview, potentially causing confusion for users.

Created a jest test file and ran a test to ensure that the css to hide head elements and scripts is being injected, here is a screenshot below

jest

This is the main change to the frame.ts file

yum

Now when adding css to the javascript and algorithm data structure beta, the scripts texts from head and body arent rendered anymore in preview

haq

### CODE REVIEW

**Comment by Othniel01** [comment]
I apologize for the mistake, The title has been changed now to match the pull request naming convention.

File: iframe-preview.test.js
Comment (naomi-lgbt) [inline iframe-preview.test.js:1]: First, this file doesn't seem to be in the right directory. Or any directory.

Second, I think rather than testing we inject the content, we should have an E2E test which enters the CSS that triggers the issue into the editor, and confirm that the preview window does not contain content.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: iframe-preview.test.js
Comment (Othniel01) [reply iframe-preview.test.js:1]: Okay I was having a hard time finding the editor element so I can insert the script into it.
This is the code in the newly created E2e test,  e2e/iframe-script.spec.ts

import { expect, test } from '@playwright/test';
import translations from '../client/i18n/locales/english/translations.json';
import { focusEditor } from './utils/editor';

test.beforeEach(async ({ page }) => {
  await page.goto(
    '/learn/javascript-algorithms-and-data-structures-v8/build-a-cash-register-project/build-a-cash-register'
  );
});

test.describe('Challenge Preview Component', () => {
  // Test for inserting <style> tag and checking if the preview is empty
  test('should render an empty preview after inserting CSS', async ({
    page,
    isMobile
  }) => {
    await focusEditor({ page, isMobile });

    // Insert the <style> tag with display: block rule into the editor
    const styleTag = `
      <style>
      * {
        margin: 0;
        padding: 0;
        display: block;
      }
      </style>`;
    await page.keyboard.insertText(styleTag);

    if (isMobile) {
      await page
        .getByRole('tab', { name: translations.learn['editor-tabs'].preview })
        .click();
    }

    // Check that the iframe preview is empty
    const frame = page.frameLocator('#fcc-main-frame').first();

    // Make sure there are no visible elements inside the iframe's body
    await expect(frame.locator('body')).toBeEmpty();
  });
});

Comment (naomi-lgbt) [reply iframe-preview.test.js:1]: The E2E test covers this change. This file should be removed.

**Comment by naomi-lgbt** [comment]
I think you misunderstood. The E2E test file should stay. The file I commented on, which is the random `iframe-preview.test.js` in the root directory, is the one that should be removed.

File: iframe-preview.test.js
Comment (Othniel01) [reply iframe-preview.test.js:1]: The e2e/iframe-script.spec.ts has been removed. If im understanding you correctly there is already another E2E test to cover this change.

hosting locally, using the problematic css does not bring up the unwanted scripts any more.

haq

**Comment by Othniel01** [comment]
yes i probably misunderstood. Please give me a second

**Comment by Othniel01** [comment]
> I think you misunderstood. The E2E test file should stay. The file I commented on, which is the random `iframe-preview.test.js` in the root directory, is the one that should be removed.

I have added the E2E test file back and removed the random unwanted jest file. 

iframe ts

**Review by naomi-lgbt — APPROVED** [review]

**Comment by Othniel01** [comment]
This checks failing seem to be coming from another file navigation-from-last-challenge.spec.ts:61:7 on chromium and chrome 
Is that an issue ?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I'm afraid this is a problem - it's no longer possible to pass the tests for the Build a Personal Portfolio Webpage project.

What's happening is that the library we use to analyse the learner's code (style code in this case) is detecting the new style element introduced by this PR.

The bigger problem is that we have a battery of tests that should have detected this, but they didn't. I'll look into that now, since we need those to work to prevent regressions.

File: client/src/templates/Challenges/utils/frame.ts [hunk 1]
```diff
@@ -88,6 +88,11 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
 
 const createHeader = (id = mainPreviewId) => `
   <base href='' />
+  <style>
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:91]:
```suggestion
  <style class="fcc-hide-header">
```
The CSSHelper script ignores `style` elements with classes, so this should fix those bugs. However, this is a bit of a hack, so let's wait until https://github.com/freeCodeCamp/freeCodeCamp/pull/56598 is in before we test this.

Assuming that this does the job, an explanatory comment (on line 87 or so) would help since it's far from obvious what this is doing.

Comment (Othniel01) [reply client/src/templates/Challenges/utils/frame.ts:91]: If this makes the style scripts ignored by the library, then for now this should be a temporal workaround preventing the unexpected rendering behavior in the test environments.

Comment (Othniel01) [reply client/src/templates/Challenges/utils/frame.ts]: Header PR has been approved, can we test this ?

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/utils/frame.ts]: Sure. I've set the tests going.

Comment (Othniel01) [reply client/src/templates/Challenges/utils/frame.ts]: 1 failing

  1) Check challenges
       Language: english
         SuperBlock: 2022/responsive-web-design
           Block: learn-html-by-building-a-cat-photo-app
             Title: Step 1
               ID: 5dc174fcf86c76b9248c6eb2
                 Check tests against solutions
                   Solution 1 must pass the tests:
     TypeError: createHeader is not a function
      at initializeTestRunner (test/test-challenges.js:671:25)
      at async getContextEvaluator (test/test-challenges.js:640:3)
      at async createTestRunner (test/test-challenges.js:591:21)
      at async Context.<anonymous> (test/test-challenges.js:539:42)



 ELIFECYCLE  Test failed. See above for more details.
 ELIFECYCLE  Command failed with exit code 1.
ERROR: "test:curriculum" exited with 1.
 ELIFECYCLE  Test failed. See above for more details.
Error: Process completed with exit code 1.

File: client/src/templates/Challenges/utils/frame.ts [hunk 2]
```diff
@@ -86,8 +86,17 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
 // of the frame.  React dom errors already appear in the console, so onerror
 // does not need to pass them on to the default error handler.
 
-export const createHeader = (id = mainPreviewId) => `
+// The "fcc-hide-header" class is added to ensure that the CSSHelper class ignores this style element
+// during tests, preventing CSS-related test failures. This is a temporary fix.
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:90]:
```suggestion
// during tests, preventing CSS-related test failures.
```
Nitpick: we don't have any real plans to change this.

File: client/src/templates/Challenges/utils/frame.ts [hunk 3]
```diff
@@ -86,8 +86,17 @@ const DOCUMENT_NOT_FOUND_ERROR = 'misc.document-notfound';
...
+
+const createHeader = (id = mainPreviewId) =>
+  `
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:93]:
```suggestion
export const createHeader = (id = mainPreviewId) => `
```
This is used by the tests, so needs to be exported.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

**Comment by ojeytonwilliams** [comment]
Hi, @Othniel01, thanks for updating the PR. It looks like the style element is causing another test to fail. Once that's addressed, this should be good to go.

**Comment by Othniel01** [comment]
Thank you @ojeytonwilliams @naomi-lgbt 
Please let me know in any way I can be of more help, as I am learning a lot from this contribution.

**Comment by ojeytonwilliams** [comment]
Thanks for the offer, @Othniel01. If you have a chance to look into the failing test, that would be a big help.

**Comment by Othniel01** [comment]
<h1> Reason for test failure </h1>
After taking time to review the failing test, seems the test failing is related to the check for a solution in the "learn-html-by-building-a-cat-photo-app" challenge from the "2022/responsive-web-design" super block.
Our injected code is making this test to fail, as the test is meant to check if any extensions are in use
<br></br>
<p>You appear to be using a browser extension that is modifying the page. Be sure to turn off all browser extensions.</p>
and return a value of +0 if successful but we are getting -1. 
<br></br>
The test framework might have detected our style injection modification and flagged the solution as invalid, even though the solution itself might be correct. 

Now we have added a csshelper class earlier that is meant to make the tests ignore our script, but it seems not all of them ignore the script.

**Comment by ojeytonwilliams** [comment]
That sounds right. It should be sufficient to adjust the test so that it checks that there is one style element (the one added in this PR). If it's anything other than 1, it's likely that an extension has added more.

File: curriculum/i18n-curriculum
Comment (ojeytonwilliams) [inline curriculum/i18n-curriculum:1]: Could you revert these changes? It's never necessary to modify the i18n-curriculum directly - that's all handled by tooling.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks for updating those tests. Could you revert the wanted changes, then we can try CI again?

**Comment by Othniel01** [comment]
<h1>Reverted Pull Request Changes</h1>
@ojeytonwilliams  Now in all the versions from english to ukrainian
<br></br>
Sorry how do we then set the length to 1, i there another way to make the change?  to give room for our injected style element, any other style element asides from this would trigger this error as an extension is in use.
<br></br>
<b>assert.equal(document.querySelectorAll('style').length, 1);</b>
<br></br>
Hopefully this stops the errors.

File: curriculum/i18n-curriculum
Comment (Othniel01) [reply curriculum/i18n-curriculum:1]: Ive reverted the changes. Is there another way to set the lengths to 1 ?

**Comment by ojeytonwilliams** [comment]
Sorry, I think we might be talking about different things.

The problem was that you'd changed the i18n curriculum. That was the only change that needed reverting.

Could you restore the fixes you had made to the curriculum files?

**Comment by Othniel01** [comment]
Okay restored back the curriculum changes and left the  i18n curriculum folder alone

**Comment by Othniel01** [comment]
Seems like we are resolving the failing tests a step at a time, the new failing test stems from the "2022/responsive-web-design" curriculum while working on the "learn-basic-css-by-building-a-cafe-menu" block, specifically for "Step 12."

"You should not add a new <style> tag. Add the new CSS rules to the existing <style> tag."

The error indicates that our code did not meet the test requirements, which expected us to modify an existing <style> tag rather than create a new one.

 1 failing

  1) Check challenges
       Language: english
         SuperBlock: 2022/responsive-web-design
           Block: learn-basic-css-by-building-a-cafe-menu
             Title: Step 12
               ID: 5f3477ae9675db8bb7655b30
                 Check tests against solutions
                   Solution 1 must pass the tests:
     Check next step for solution!
Test text: <p>You should not add a new <code>style</code> tag. Add the new CSS rules to the existing <code>style</code> tag.</p>
  AssertionError: Unspecified AssertionError
      at eval (eval at <anonymous> (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:123872), <anonymous>:2:1)
      at HTMLDocument.<anonymous> (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:123872)
      at u (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:26888)
      at c (http://127.0.0.1:8080/js/frame-runner-543042f22803f54627ea.js:2:27190)

**Comment by Othniel01** [comment]
This is the code responsible for checking if there are more than 1 style sheets added or being used at the "2022/responsive-web-design" curriculum while working on the "learn-basic-css-by-building-a-cafe-menu" block, specifically for "Step 12."

const hasManyStyleTags = document.querySelectorAll('style').length > 1;
assert(!hasManyStyleTags);


Since we are injecting a style sheet globally, we can increase the number to 2

const hasManyStyleTags = document.querySelectorAll('style').length > 2;
assert(!hasManyStyleTags);

Allowing for our injected style script and the given style script and not more than two. any other style added would now throw the error ignoring our injected style sheet

File: curriculum/challenges/english/25-front-end-development/workshop-cafe-menu/5f3477ae9675db8bb7655b30.md [hunk 4]
```diff
@@ -21,7 +21,7 @@ assert(hasH1);
 You should not add a new `style` tag. Add the new CSS rules to the existing `style` tag.
 
 ```js
-const hasManyStyleTags = document.querySelectorAll('style').length > 1;
+const hasManyStyleTags = document.querySelectorAll('style').length > 2;
 assert(!hasManyStyleTags);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/workshop-cafe-menu/5f3477ae9675db8bb7655b30.md:25]:
```suggestion
assert.isAtMost(document.querySelectorAll('style').length, 2);
```

Comment (Othniel01) [reply curriculum/challenges/english/25-front-end-development/workshop-cafe-menu/5f3477ae9675db8bb7655b30.md]: Suggested changes have been taken in to account, thank you for your cleaner, and more simplified approach.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hi, @Othniel01 I'm afraid there are more tests that were written assuming there would be one style element.

At a glance they seem to be quite brittle, so it's possible there are better ways to write them. I'll ask someone on the curriculum team to advise.

**Comment by Othniel01** [comment]
Hey @ojeytonwilliams, thanks for your feedback, I would be waiting for any updates concerning this issue

**Comment by ilenia-magoni** [comment]
there are merge conflicts now

**Comment by Othniel01** [comment]
hope the merge conflicts have been resolved now

**Comment by ilenia-magoni** [comment]
there are still challenges failing for the new style, @ojeytonwilliams do you have updates?

**Comment by ojeytonwilliams** [comment]
Sorry for the delay. I pushed a couple of minor changes to fix the failing tests and it should be fine now.

**Review by ojeytonwilliams — APPROVED** [review]
Thanks for your patience, @Othniel01. This LGTM :+1:

**Comment by ojeytonwilliams** [comment]
@ilenia-magoni it shouldn't be necessary. Every time you push a new commit to GitHub, it implicitly rebases the entire branch on top of the target, main in this case.

**Comment by ilenia-magoni** [comment]
I did not mean merge conflicts, I meant to see if new challenges have issues, or things like that

**Comment by ojeytonwilliams** [comment]
Right, but that's what just happened. When I pushed my commit, GitHub rebased this PR's commits on top of main, before triggering the actions. Any new challenges that were in main at the point I pushed will have been included in the CI tests.
